A new interface accommodating in-process backends

facundominguez opened the pull request.

qaristote committed:
split library into modules

qaristote committed:
make Solver type into a class

qaristote committed:
enable and fix warnings

qaristote committed:
solver: backend: fix comment

qaristote committed:
use bytestrings to communicate with backends

qaristote committed:
have backends send lazy bytestrings instead of strict ones

qaristote committed:
add Z3 solver backend

qaristote committed:
remove extra whitespace

qaristote committed:
remove extra whitespace

qaristote committed:
remove duplicated file

qaristote committed:
solver: backend: use a record instead of a typeclass

qaristote committed:
restructure repository

qaristote committed:
solver: remove the loadString function

its sole purpose is to check that the input string is well-formatted,
but this is already dealt with by the backends
the send function should be used instead

qaristote committed:
solver: add the option for laziness (the commands are only sent to the solver when necessary)

qaristote committed:
print success in eager mode, not the converse

qaristote committed:
add some comments

qaristote committed:
pull parsing back to the common interface

qaristote committed:
pull back logging to the common interface

qaristote committed:
backends: process: rewrite to use the more recent typed-process and async libraries

qaristote committed:
make the interface of backends more uniform

qaristote committed:
improve the interface of the backends

qaristote committed:
improve the interface of the backends

facundominguez commented on src/SimpleSMT/Solver.hs:
```diff
@@ -0,0 +1,395 @@
+  }
+
+-- | Send a command to the solver.
+sendSolver :: Solver -> LBS.ByteString -> IO ()
```
Does the function return only after the solver produces a response? Otherwise, how do we know when `recvSolver` is good to call?

facundominguez replied:
I'm wondering what is the motivation to have `sendSolver` and `recvSolver`. I would have expected something like:
```
sendSolver :: Solver -> LBS.ByteString -> IO Sexp

-- | Like 'sendSolver_' but discards the result.
sendSolver_ :: Solver -> LBS.ByteString -> IO ()
```

qaristote replied:
> Does the function return only after the solver produces a response?

This is true for the `Z3` backend and I expect it to be true for the `Process` backend as well because [this is how simple-smt works internally](https://github.com/yav/simple-smt/blob/dd03ca119ddcd50bde52fbe40c96eb60cc1d6338/SimpleSMT.hs#L319-L321) (as far as I understand). I agree that this is not ideal, and I had planned to say a word about this in the documentation.

qaristote replied:
As to why I did this, here's my reasoning. Before making this change the interface provided a single
```
send :: Solver -> LBS.ByteString -> IO SExpr
```
I'd rather the output type be `LBS.ByteString` as well, so that parsing and logging the output can be taken care of in the common interface. 

The problem with that is that it's not really compatible with the `Process` backend. Since you cannot assume that the solver writes its output on a single-line (Z3 does not and I didn't find a way to configure it to do so), your only option to read the output is to use `hGetContents` and parse it until you get a full s-expression. So if you wanted to return an `LBS.ByteString` you'd have to parse the bytestring as an s-expression, serialize it back to a bytestring, only for it to be parsed again in the common interface.

My solution to that is to use this `response :: IORef ByteString` buffer, and the `sendSolver` and `recvSolver` functions that are deduced from it. To the user this shouldn't change anything because `command` is just `sendSolver >> recvSolver`, and this also has the advantage that parsing the `(success)`s when sending batch of commands is also implementable.

qaristote replied:
A solution to have
```haskell
send :: Solver -> LBS.ByteString -> IO LBS.ByteString
```
would be to reimplement [this](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/src/Data.ByteString.html#hGetContentsSizeHint) for the `Process` backend.

qaristote replied:
Oh I just realized was probably just about what functions are available in the API ? `sendSolver` and `recvSolver` aren't currently exposed in the API anyways, I just separated them because I thought it was cleaner but it's not particularly necessary.

facundominguez replied:
Your explanation makes sense to me. I think, though, that maybe it is possible to deal with the output of z3 as discussed in #2.

facundominguez commented on Z3/src/SimpleSMT/Solver/Z3.hs:
```diff
@@ -0,0 +1,90 @@
+      -- Z3 requires null-terminated cstrings
+      -- appending the null character performs a memcpy so is inefficient
+      -- TODO a better solution would be to do this on the bytestring-build before it
+      -- is evaluated to a lazy bytestring
```
This corresponds to changing the interface of `sendSolver` to demand a NUL-terminated bytestring.

Another solution would be to have `sendSolver` receive a Builder instead of a bytestring, in which case appending a null would be cheaper.

I think that an extra `memcpy` would happen in `toStrict` as a result of appending a NUL character if `cmd` contains at most one chunk. Otherwise, a memcpy would occur regardless of whether the NUL character is appended.

qaristote replied:
> This corresponds to changing the interface of sendSolver to demand a NUL-terminated bytestring.

Yeah I didn't want to do that because this is a Z3-specific problem so it doesn't make sense to enforce that to all builders (this would force the `Process` builder to drop the last character for example).

> Another solution would be to have sendSolver receive a Builder instead of a bytestring, in which case appending a null would be cheaper.

That's what I was thinking of, especially since it also allows backends to specify their own building strategy (`safeStrategy` or `untrimmedStrategy`). But this means both single commands and batches of commands need to be built with the same chunk size (unless the chunk sizes are also passed as arguments, but this would be awkward in the API).

Another solution would be to add the final character (and also the strategy for that matter) as a field in the `Backend` datatype.

> I think that an extra memcpy would happen in toStrict as a result of appending a NUL character if cmd contains at most one chunk. Otherwise, a memcpy would occur regardless of whether the NUL character is appended.

Yeah this is definitely not a pressing matter hence why I'm leaving this as-is at least for now.

qaristote replied:
Havin `send` receive a `Builder` would also have the advantage that the `Process` backend could then directly use the more efficient `hPutBuilder`.

facundominguez replied:
> But this means both single commands and batches of commands need to be built with the same chunk size (unless the chunk sizes are also passed as arguments, but this would be awkward in the API).

I think I don't understand what is the need to use different chunk sizes in these scenarios.

qaristote replied:
It's just a matter of optimization:

> A sub-performing case of [toLazyByteString](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#v:toLazyByteString) is executing short (<128 bytes) [Builder](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#t:Builder)s. In this case, the allocation overhead for the first 4kb buffer and the trimming cost dominate the cost of executing the [Builder](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#t:Builder).

IIRC it made a noticeable difference in Pirouette. But if all backends use the `untrimmedStrategy` then that cost is alleviated.

qaristote replied:
https://github.com/tweag/simple-smt/pull/1/commits/3355b1227decba76cda050c6b23b5ec74ad16605 moves the evaluation of the builders inside the backends.

qaristote committed:
SExpr: serialization: use untrimmedStrategy instead of safeStrategy

qaristote committed:
solver: get rid of the response buffer

qaristote committed:
have Backend.send take a bytestring builder as input

this allows the process backend to directly put the builder on a pipe instead of evaluating it first

qaristote committed:
Merge branch 'master' into qa/process_hGetContentsEager

qaristote committed:
backends: process: do some basic parsing to extract the next command

qaristote committed:
backends: process: toBackend: replace lazy IO with repeated line reading, and improve parsing

qaristote committed:
add test suite

qaristote committed:
tests: check parsed value against an s-expr instead of a string

qaristote committed:
tests: SExpr: also test bytestring parsing

qaristote committed:
tests: enable warnings and fix them

qaristote committed:
backends: rewrite with using Control.Exception.bracket

qaristote committed:
sexpr: fix bytestring parser

qaristote committed:
Merge branch 'master' into tests

qaristote committed:
tests: sexpr: test parsing of complex terms

qaristote committed:
sexpr: parsing: add case for numerals

qaristote committed:
Merge branch 'qa/fix_parsing' into tests

qaristote committed:
SExpr: parsing: fix handling of "\# in parsing of string literals

qaristote committed:
Merge branch 'qa/fix_parsing' into tests

qaristote committed:
tests: sources: fix 'terms' source

qaristote committed:
tests: restructure code

qaristote committed:
(wip) tests: sources: add computation associated to each example

qaristote committed:
tests: sources: add associated computation to tests

qaristote committed:
Merge branch 'master' into qa/process_hGetContentsEager

qaristote committed:
Merge pull request #2 from tweag/qa/process_hGetContentsEager

Get rid of the response buffer

qaristote committed:
Merge branch 'master' into tests

qaristote committed:
tests: add solver tests

qaristote committed:
tests: solver: process: fix tests and process backend

qaristote committed:
remove extra whitespace

qaristote committed:
tests: expose reusable parts

qaristote committed:
Z3: add tests

qaristote committed:
tests: move module into its own library

facundominguez commented on src/SimpleSMT/Solver.hs:
```diff
@@ -0,0 +1,383 @@
+
+-- | A command with no interesting result.
+-- In eager mode, the result is checked for correctness.
+-- In lazy mode, (unless the queue is flushed and evaluated
```
Are eager and lazy modes explained somewhere?

qaristote replied:
No that's all there is, writing better documentation is on my to-do list ^^

qaristote replied:
https://github.com/tweag/simple-smt/pull/5

facundominguez commented on src/SimpleSMT/Solver.hs:
```diff
@@ -0,0 +1,383 @@
+  sendSolver solver =<<
+    case queue solver of
+      Nothing -> return $ cmd
+      Just q -> (<> renderSExpr expr) <$> flushQueue q
```
How about reusing the logic of queues for commands taking `SExpr` and commands taking `Builder`?

qaristote replied:
I'm not sure I understand what you mean, could you develop ?

facundominguez replied:
Sure. As a user of this library, which function should I call to send commands to the SMT solver if I want to have queuing of commands but I want to provide a `Builder` and not a `SExpr`?

Does that clarify it?

qaristote replied:
Yes I get your point. Currently there's no way to do that, as I was assuming the user would only be writing `SExpr`s. But it makes sense to make it possible indeed.

qaristote committed:
add documentation for the Solver module

this describes in particular the eager and lazy mode more in depth

qaristote committed:
doc: solver: add usage example

qaristote committed:
doc: sexpr: explain readSExpr is currently incorrect

qaristote committed:
doc: backends: add short module descriptions

qaristote committed:
doc: add documentation the to part of the test suite that's exposed

qaristote committed:
process backend: set process handles to binary mode

gabrielhdt committed:
Add action checking ormolu style

qaristote committed:
process backend: import hSetBinaryMode

qaristote committed:
doc: solver: initSolverWith: add a link to Solver

Co-authored-by: Facundo Domínguez <facundominguez@gmail.com>

qaristote committed:
doc: solver: reduce max line size

qaristote committed:
Merge pull request #5 from tweag/doc

improve documentation

qaristote committed:
process: handles: force block buffering

qaristote committed:
Merge pull request #7 from tweag/qa/process/set_handle_options

`Solver.Process`: set handle options

gabrielhdt committed:
Check all files

gabrielhdt committed:
Format them all

gabrielhdt committed:
Don't blame Ormolu ;(

qaristote committed:
Merge pull request #6 from tweag/auto-format

use ormolu to format the code automatically

gabrielhdt committed:
Add Nix flake and HLS setup (#9)

gabrielhdt committed:
Run tests in CI (#10)

* Run tests in CI

* Ignoring process backend tests

gabrielhdt committed:
Do not test readSExpr (#12)
